src/placeos-models/user.cr

```diff
@@ -85,11 +86,48 @@ module PlaceOS::Model
     # Callbacks
     ###############################################################################################
 
+    # Locking to protect against concurrent deletes
+
+    before_destroy :lock_admin_deletes
+    after_destroy :unlock_admin_deletes
+
+    before_destroy :ensure_admin_remains
     before_destroy :cleanup_auth_tokens
     before_save :build_name
     before_save :create_email_digest
 
-    # deletes auth tokens
+    private getter admin_destroy_lock : RethinkORM::Lock do
+      RethinkORM::Lock.new("admin_destroy_lock")
+    end
+
+    # :inherit:
+    def destroy
+      super
+    ensure
+      admin_destroy_lock.unlock if admin_destroy_lock.locked?
+    end
+
+    protected def lock_admin_deletes
+      return unless self.sys_admin
+      admin_destroy_lock.lock
+    end
+
+    protected def unlock_admin_deletes
+      return unless self.sys_admin
+      admin_destroy_lock.unlock if admin_destroy_lock.locked?
+    end
+
+    # Prevent the system from entering a state with no admin
+    protected def ensure_admin_remains
+      return unless self.sys_admin
+
+      if User.count(sys_admin: true) == 1
+        admin_destroy_lock.unlock if admin_destroy_lock.locked?
+        raise Model::Error.new("At least one admin must remain")
+      end
+    end
```

Comment:
To reduce some of that manual lock management, what about?
```suggestion
    def destroy
      return super unless self.sys_admin
      
      admin_destroy_lock.synchronize do
        if User.count(sys_admin: true) == 1
          raise Model::Error.new("At least one admin must remain")
        end
        super
      end
    end
```

Reply:
I would prefer to keep the actual check within a callback, however, the synchronisation in the overridden `destroy` is definitely simpler (despite duplicating the sys_admin check)

Reply:
Something along the lines of `around_*` callbacks would be useful to prevent patching orm methods.

Reply:
Agreed - had a look for exactly that!